dask/base.py

```diff
@@ -377,36 +379,69 @@ def optimization_function(x):
     return getattr(x, "__dask_optimize__", dont_optimize)
 
 
-def collections_to_dsk(collections, optimize_graph=True, optimizations=(), **kwargs):
+def newstyle_collections(collections):
+    from dask.delayed import Delayed
+
+    is_newstyle = [
+        isinstance(c, DaskCollection2) and not isinstance(c, Delayed)
+        for c in collections
+    ]
+    if any(is_newstyle):
+        if not all(is_newstyle):
+            raise RuntimeError("Provided new- and old-style collections.")
+        if not all(
+            isinstance(c.__dask_graph_factory__(), TaskGraphFactory)
+            for c in collections
+        ):
+            raise TypeError("Newstyle collections must have a TaskGraphFactory graph.")
+
+        return True
+    return False
+
+
+def collections_to_dsk(
+    collections, optimize_graph=True, optimizations=(), **kwargs
+) -> TaskGraphFactory:
     """
     Convert many collections into a single dask graph, after optimization
     """
-    from dask.highlevelgraph import HighLevelGraph
-
-    optimizations = tuple(optimizations) + tuple(config.get("optimizations", ()))
-
-    if optimize_graph:
-        groups = groupby(optimization_function, collections)
+    from dask.highlevelgraph import HighLevelGraph, TaskFactoryHLGWrapper
 
-        graphs = []
-        for opt, val in groups.items():
-            dsk, keys = _extract_graph_and_keys(val)
-            dsk = opt(dsk, keys, **kwargs)
-
-            for opt_inner in optimizations:
-                dsk = opt_inner(dsk, keys, **kwargs)
-
-            graphs.append(dsk)
-
-        # Merge all graphs
-        if any(isinstance(graph, HighLevelGraph) for graph in graphs):
-            dsk = HighLevelGraph.merge(*graphs)
+    if newstyle_collections(collections):
+        graph_factories = [c.__dask_graph_factory__() for c in collections]
+        if len(graph_factories) > 1:
+            expr = type(graph_factories[0]).combine_factories(*graph_factories)
         else:
-            dsk = merge(*map(ensure_dict, graphs))
+            expr = collections[0]
+        return expr.optimize()
     else:
-        dsk, _ = _extract_graph_and_keys(collections)
-
-    return dsk
+        optimizations = tuple(optimizations) + tuple(config.get("optimizations", ()))
+        if optimize_graph:
+            ext_keys = []
+            groups = groupby(optimization_function, collections)
+
+            graphs = []
+            for opt, val in groups.items():
+                dsk, keys = _extract_graph_and_keys(val)
+                ext_keys.extend(keys)
+                dsk = opt(dsk, keys, **kwargs)
+
+                for opt_inner in optimizations:
+                    dsk = opt_inner(dsk, keys, **kwargs)
+
+                graphs.append(dsk)
+
+            # Merge all graphs
+            if any(isinstance(graph, HighLevelGraph) for graph in graphs):
+                dsk = HighLevelGraph.merge(*graphs)
+                return TaskFactoryHLGWrapper(dsk, ext_keys)
```

Comment:
I defined this wrapper just to streamline stuff in distributed. As pointed out in https://github.com/dask/distributed/pull/7942#issuecomment-1864651937 we need more than just the graph, we also need the output keys to actually know what to compute